src/labCommands.ts

```diff
@@ -0,0 +1,457 @@
+          const { content } = current;
+          if (content.activeCell !== null) {
+            const editor = content.activeCell.editor as CodeMirrorEditor;
+            (CodeMirror as any).Vim.handleKey(editor.editor, '<Esc>');
+          }
+        }
+      },
+      isEnabled
+    }),
+    commands.addCommand('select-below-execute-markdown', {
+      label: 'Execute Markdown and Select Cell Below',
+      execute: args => {
+        const current = getCurrent(args);
+
+        if (current) {
+          const { content } = current;
+          if (
+            content.activeCell !== null &&
+            content.activeCell.model.type === 'markdown'
+          ) {
+            (current.content.activeCell as MarkdownCell).rendered = true;
+          }
+          return NotebookActions.selectBelow(current.content);
+        }
+      },
+      isEnabled
+    }),
+    commands.addCommand('select-above-execute-markdown', {
+      label: 'Execute Markdown and Select Cell Below',
+      execute: args => {
+        const current = getCurrent(args);
+
+        if (current) {
+          const { content } = current;
+          if (
+            content.activeCell !== null &&
+            content.activeCell.model.type === 'markdown'
+          ) {
+            (current.content.activeCell as MarkdownCell).rendered = true;
+          }
+          return NotebookActions.selectAbove(current.content);
+        }
+      },
+      isEnabled
+    }),
+    commands.addCommand('select-first-cell', {
+      label: 'Select First Cell',
+      execute: args => {
+        const current = getCurrent(args);
+
+        if (current) {
+          const { content } = current;
+          content.activeCellIndex = 0;
+          content.deselectAll();
+          if (content.activeCell !== null) {
+            ElementExt.scrollIntoViewIfNeeded(
+              content.node,
+              content.activeCell.node
+            );
+          }
+        }
+      },
+      isEnabled
+    }),
+    commands.addCommand('select-last-cell', {
+      label: 'Select Last Cell',
+      execute: args => {
+        const current = getCurrent(args);
+
+        if (current) {
+          const { content } = current;
+          content.activeCellIndex = current.content.widgets.length - 1;
+          content.deselectAll();
+          if (content.activeCell !== null) {
+            ElementExt.scrollIntoViewIfNeeded(
+              content.node,
+              content.activeCell.node
+            );
+          }
+        }
+      },
+      isEnabled
+    }),
+    commands.addCommand('center-cell', {
+      label: 'Center Cell',
+      execute: args => {
+        const current = getCurrent(args);
+
+        if (current && current.content.activeCell !== null) {
+          const er = current.content.activeCell.inputArea.node.getBoundingClientRect();
+          current.content.scrollToPosition(er.bottom, 0);
+        }
+      },
+      isEnabled
+    }),
+
+    commands.addKeyBinding({
```

Comment:
It would be nice to define those in the settings (see [documentation](https://jupyterlab.readthedocs.io/en/stable/extension/extension_points.html#keyboard-shortcuts)) to allow user customization.

Reply:
This has definitely been requested before and I agree it would be good. But it doesn't seem that that would allow easily disposing of them when we turn vim mode off?

Reply:
Would the way be to:
1. Add all our commands to the registry
2. Set the default shortcuts in the settings
3. disable our commands when vim mode is active

Reply:
ok let's include that here. This is the natural place for that

Reply:
You would recommend adding a data attribute to the notebook (when the vim mode is activate) and to add it to the key binding selector. 

See for example: https://github.com/jupyterlab/jupyterlab/blob/6871e89ff4ab8cfd32c59acdfb6fd60e4a9018a3/packages/console-extension/schema/tracker.json#L76

The reason is that the `isEnabled` is tested for after looking for the selector match. So better removing it through the selector directly.

Reply:
Ah that is slick, but it seems that would also require us to start messing around with css. Is there an easy way for apply that? It seems that the console is using the fact that it's also responsible for creating these instrances in order to keep track of them: 

https://github.com/jupyterlab/jupyterlab/blob/6871e89ff4ab8cfd32c59acdfb6fd60e4a9018a3/packages/console-extension/src/index.ts#L377-L405

Reply:
ah nevermind I've figured it out